source/loader/layers/sanitizer/msan/msan_interceptor.cpp

```diff
@@ -0,0 +1,493 @@
+        &DI->Alignment, nullptr));
+
+    // Don't move DI, since it's a return value as well
+    m_DeviceMap.emplace(Device, DI);
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::eraseDevice(ur_device_handle_t Device) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_DeviceMapMutex);
+    assert(m_DeviceMap.find(Device) != m_DeviceMap.end());
+    m_DeviceMap.erase(Device);
+    // TODO: Remove devices in each context
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::insertProgram(ur_program_handle_t Program) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_ProgramMapMutex);
+    if (m_ProgramMap.find(Program) != m_ProgramMap.end()) {
+        return UR_RESULT_SUCCESS;
+    }
+    m_ProgramMap.emplace(Program, std::make_shared<ProgramInfo>(Program));
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::eraseProgram(ur_program_handle_t Program) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_ProgramMapMutex);
+    assert(m_ProgramMap.find(Program) != m_ProgramMap.end());
+    m_ProgramMap.erase(Program);
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::insertKernel(ur_kernel_handle_t Kernel) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
+    if (m_KernelMap.find(Kernel) != m_KernelMap.end()) {
+        return UR_RESULT_SUCCESS;
+    }
+    m_KernelMap.emplace(Kernel, std::make_shared<KernelInfo>(Kernel));
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::eraseKernel(ur_kernel_handle_t Kernel) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
+    assert(m_KernelMap.find(Kernel) != m_KernelMap.end());
+    m_KernelMap.erase(Kernel);
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t
+MsanInterceptor::insertMemBuffer(std::shared_ptr<MemBuffer> MemBuffer) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_MemBufferMapMutex);
+    assert(m_MemBufferMap.find(ur_cast<ur_mem_handle_t>(MemBuffer.get())) ==
+           m_MemBufferMap.end());
+    m_MemBufferMap.emplace(reinterpret_cast<ur_mem_handle_t>(MemBuffer.get()),
+                           MemBuffer);
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::eraseMemBuffer(ur_mem_handle_t MemHandle) {
+    std::scoped_lock<ur_shared_mutex> Guard(m_MemBufferMapMutex);
+    assert(m_MemBufferMap.find(MemHandle) != m_MemBufferMap.end());
+    m_MemBufferMap.erase(MemHandle);
+    return UR_RESULT_SUCCESS;
+}
+
+std::shared_ptr<MemBuffer>
+MsanInterceptor::getMemBuffer(ur_mem_handle_t MemHandle) {
+    std::shared_lock<ur_shared_mutex> Guard(m_MemBufferMapMutex);
+    if (m_MemBufferMap.find(MemHandle) != m_MemBufferMap.end()) {
+        return m_MemBufferMap[MemHandle];
+    }
+    return nullptr;
+}
+
+ur_result_t MsanInterceptor::prepareLaunch(
+    std::shared_ptr<DeviceInfo> &DeviceInfo, ur_queue_handle_t Queue,
+    ur_kernel_handle_t Kernel, USMLaunchInfo &LaunchInfo) {
+    auto Program = GetProgram(Kernel);
+
+    do {
```

Comment:
please remove this loop.
I also noticed that the same thing is done in the same asan function. Please remove that as well.